docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
+- Support for Helm install, upgrade, test, rollback and uninstall
+- Performs Helm uninstall on `HelmRelease` CR removal
+- Chart values override\
+- Pull charts from Helm repository directly
+
+_Develop our own Homebrew Helm Controller_
+
+This approach affords us with a controller that meets our specific needs and avoids any third party software dependency. 
+
+
+
+
+## Status
+
+Proposed
+
+## Open Questions
```

Comment:
IMO we need to answer more less the following questions to be able to know if we should or NOT adopt flux:

**Complexity:**

- While FluxCD provides a powerful feature set, configuring and using it can be complex. How much effort will be required to integrate and become proficient in Helm + Flux?
- Considering the extensive features offered by Flux, do we truly need all of them, or would only a subset be sufficient? Will the integration justify the learning curve and potential limitations?
- How complex will it be to test the solution with this dependency in local development and CI environments?

**Compatibility:**
- Which Kubernetes versions are supported by Flux? Are there any limitations?
- What are the supported Helm versions and the extent of Helm features that Flux can handle? 
- Can Flux handle Helm hooks effectively?
- How does Flux manage its release lifecycle, and what are the support policies?

**Automation Limitations**

Despite its automation capabilities, are there any aspects of the deployment process that Flux may not be able to automate effectively, requiring manual intervention or additional tooling. Therefore, will adopting Flux truly eliminate the need to use Helm as a dependency for functions such as installation, upgrade, uninstallation, or checking the status of a release? Or will we end up with both implementations, adding unnecessary complexities?

**Flux as a Dependency:**

- Can Flux be effectively used as a dependency within the Operator? Is it possible to create a Flux CR from the Operator and interact with Flux's internal functionality? Or are not they allowing importing their APIs? Is their code implementation exported for usage or not?

**What happens when we need bug fixes and RFEs if we adopt flux?

- If we face issues within the project, will Flux provide support for such cases, and if so, what areas do they support?
- In the event that we need a feature enhancement or bug fix in Flux to address our specific needs, how will that process work?

**Flux Features**
- Which specific features provided by Flux would be advantageous to adopt in our project?
- What are our requirements that would be fulfilled by adopting Flux? In other words, what functionalities would we not need to implement separately due to Flux integration?
- Are there any scenarios where using Flux could create limitations or challenges in managing HelmCharts or maintaining control over them?